lib/src/over_react_redux/hooks/use_selector.dart

```diff
@@ -0,0 +1,183 @@
+          equalityFn(unwrapInteropValue(nextJsSelector), unwrapInteropValue(prevJsSelector)));
+
+  return unwrapInteropValue<TSelector>(_jsUseSelector(allowInterop(jsSelector), jsEqualityFn));
+}
+
+@JS('ReactRedux.useSelector')
+external ReactInteropValue _jsUseSelector(_JsSelectorFn selector, [_JsReduxStateEqualityFn equalityFn]);
+
+// ----------------------------------------------------
+//  createSelectorHook
+// ----------------------------------------------------
+
+/// Returns a function that can be named and used as a replacement for [useSelector] when you want to
+/// use custom [context].
+///
+/// This is useful if you're building a complex reusable component, and you don't want your `Store` to collide
+/// with any Redux `Store` your consumers' applications might use.
+///
+/// See the [react-redux JS documentation](https://react-redux.js.org/api/hooks#custom-context) for more details.
+///
+/// ### Custom Context Example
+///
+/// ```dart
+/// import 'package:over_react/over_react.dart';
+/// import 'package:over_react/over_react_redux.dart';
+/// import 'package:redux/redux.dart';
+///
+/// // ------------------------------------
+/// //  1. Declare the custom context
+/// // ------------------------------------
+/// final MyContext = createContext<MyState>();
+/// final useMySelector = createSelectorHook(MyContext);
+/// final myStore = Store(myReducer);
+///
+/// // ------------------------------------
+/// //  2. (Optional) Create a custom
+/// //  provider component that uses the
+/// //  context defined above.
+/// // ------------------------------------
+/// mixin MyProviderProps = UiProps with ReduxProviderProps;
+///
+/// UiFactory<MyProviderProps> MyProvider = uiFunction(
+///   (props) {
+///     return (ReduxProvider()
+///       ..addUnconsumedProps(props, const [])
+///       ..context = MyContext
+///       ..store = myStore
+///     )(
+///       props.children,
+///     );
+///   },
+///   $MyProviderConfig, // ignore: undefined_identifier
+/// );
+///
+/// // ------------------------------------
+/// //  3. Create a function component that
+/// //  uses `useMySelector`.
+/// // ------------------------------------
+/// mixin MyComponentProps on UiProps {}
+///
+/// UiFactory<MyComponentProps> MyComponent = uiFunction(
+///   (props) {
+///     final count = useMySelector<int, MyState>((state) => state.count);
+///
+///     return Dom.div()('The current count is $count');
+///   },
+///   $MyComponentConfig, // ignore: undefined_identifier
+/// );
+///
+/// // ------------------------------------
+/// //  4. Render the function component
+/// //  nested within the custom provider.
+/// // ------------------------------------
+/// main() {
+///   final app = MyProvider()(
+///     MyComponent()(),
+///   );
+///
+///   react_dom.render(app, querySelector('#id_of_mount_node'));
+/// }
+/// ```
+_SelectorFnHook<TReduxState> createSelectorHook<TReduxState>([Context<TReduxState> context]) {
```

Comment:
It is nicer, but there is the extra step of creating the hook, whereas the `useSelector` hook is a "catchall" for the default `Context`.

Reply:
It also runs the risk of them exporting the shadowed entity unintentionally, which could create confusion for consumers.

Reply:
@greglittlefield-wf I tried removing the generic param from `Context` as you suggested above - but it makes it so that the typing is not inferred correctly for consumers.

<img width="631" alt="Screen Shot 2021-01-22 at 7 36 26 AM" src="https://user-images.githubusercontent.com/1750797/105504059-8e6eda00-5c84-11eb-90b6-1dc3d19f71fa.png">

Reply:
> It is nicer, but there is the extra step of creating the hook, whereas the useSelector hook is a "catchall" for the default Context.

Yeah, I was thinking it'd be worth it though in a lot of cases to take that extra step. For instance:
```dart
final useMySelector = useSelector<MyState>();

... = uiFunction(props) {
  final foo = useMySelector((state) => state.foo);
  final bar = useMySelector((state) => state.bar);
  final baz = useMySelector((state) => state.baz);
}...
```
vs
```dart
... = uiFunction(props) {
  final foo = useSelector<MyState, int>((state) => state.foo);
  final bar = useSelector<MyState, String>((state) => state.bar);
  final baz = useSelector<MyState, Baz>((state) => state.baz);
}...
```

I think we should at call it out prominently as an option to help avoid having to specifying the types.

Reply:
> It also runs the risk of them exporting the shadowed entity unintentionally, which could create confusion for consumers.
Yeah, that's fair; I suppose they can decide to name it whatever they want.


>  I tried removing the generic param from Context as you suggested above - but it makes it so that the typing is not inferred correctly for consumers.

Did you retain the generic parameter in the method and explicitly add it to the call site (since it no longer gets inferred from the context generic parameters)?
```diff
-final useMySelector = createSelectorHook(MyContext);
+final useMySelector = createSelectorHook<MyState>(MyContext);
```

> Since Redux uses the redux Store.state as the value for Context.Provider.value, it seems like the generic arg should probably be something like this:

My understanding is that Redux stores the store itself in context, not the state. See https://react-redux.js.org/api/provider#provider for more info

And like I mentioned above, it's not the Dart store that's stored directly in the context. I thought it was the JS store, but it looks like it's actually an internal Redux object wrapped around the JS store. Take this example:

```dart
      (ReduxProvider()
        ..store = store
        ..context = MyContext
      )(
      MyContext.Consumer()((value) {
        window.console.log('context value');
        window.console.log(value);
      }),
```
<img width="542" alt="Screen Shot 2021-01-22 at 9 25 30 AM" src="https://user-images.githubusercontent.com/1713967/105517187-f8db4680-5c93-11eb-8beb-d9b9391ac071.png">

So, typing it as the Dart store would be incorrect and lead to potential runtime type errors.

Reply:
> Did you retain the generic parameter in the method and explicitly add it to the call site (since it no longer gets inferred from the context generic parameters)?

Nope... that works.